refactor integration tests (#6168)
bbc/simorgh: src/integration/README.md

## How to write tests

We need to write tests for both our AMP and canonical platforms. There are tests that can be written to test both platforms (cross platform tests), some tests for canonical only, and some tests for AMP only.

#### Cross platform tests

For tests that can be run in both platforms, we should add these tests to a file called `crossPlatformTests.js`. An example of a cross platform test would be like

```js
export default () => {
  it('I can see the headline', () => {
    const headlineEl = document.getByText(headlineText);

    expect(headlineEl).toBeInTheDocument();
  });
};
```

#### AMP only tests

> For tests that should be run only on the AMP platform, we should add these tests to a file called `ampTests.js`. An example of a cross platform test would be like

Reviewer: ```suggestion
For tests that should be run only on the AMP platform, we should add these tests to a file called `ampTests.js`. An example of an AMP test would be like
```

```js
export default () => {
  describe('Analytics', () => {
    it('ATI', () => {
      expect(
        document.querySelector('amp-analytics script[type="application/json"]')
          .textContent,
      ).toMatch('https://logws1363.ati-host.net?');
    });
  });
};
```

#### Canonical only tests

> For tests that should be run only on the canonical platform, we should add these tests to a file called `canonicalTests.js`. An example of a cross platform test would be like

Reviewer: ```suggestion
For tests that should be run only on the canonical platform, we should add these tests to a file called `canonicalTests.js`. An example of a canonical test would be like
```

Before our tests run, the test environment [setup file](https://github.com/bbc/simorgh/tree/latest/src/integration/integrationTestEnvironment.js) parses the `pathname` dockblock pragma and constructs the url. JSDOM then visits the url to get the DOM trees that we can use to run our tests against.

Note we have also specified a `service` docblock pragma. The service is parsed and added to the global scope of every test file. This can come in handy for tests where you need to know the service. There are other useful variables added to the global scope:

- `pageType` - The page type which is parsed from the test file path.
- `service` - The service of the currently running page test.
- `document` - The `document` object of the current page. You can use this to query the DOM and assert the correct things are in the page.

> - `platform` - The current platform (`amp` or `canonical`)

Reviewer: Is there any need to have this in the global scope? Isn't that implicit knowledge based on whether you are writing and AMP or canonical test? And I guess a cross platform test shouldn't matter. 

I'm just wary of cross-platform tests containing `if(platform === 'amp')` rather than tests being correctly put in their respective platform related files.

Author: that's a fair point!

## How to write tests

We need to write tests for both our AMP and canonical platforms. There are tests that can be written to test both platforms (cross platform tests), some tests for canonical only, and some tests for AMP only.

#### Cross platform tests

For tests that can be run in both platforms, we should add these tests to a file called `crossPlatformTests.js`. An example of a cross platform test would be like

```js
export default () => {
  it('I can see the headline', () => {
    const h1El = document.querySelector('h1');

    expect(h1El).toBeInTheDocument();

>     expect(h1El).toBeTruthy();

Reviewer 2: Should this be `expect(h1El.textContent).toBeTruthy();` (just looking at other patterns across the tests, and we seem to check that the text, rather than the component itself is truthy

```
// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`I can see the headline`] = `
<h1 id="content"
       "tabindex="-1"
>
  <span role="text">
    <span lang="en-GB">
      BBC News
    </span>
    ,
    Pidgin
    -
    Home
  </span>
</h1>
`;
```

It's recommended not to snapshot large parts of the DOM because this creates brittle tests i.e. these tests would break often because the DOM changes whenever someone changes something such as a React component's structure and styles, translation configs and feature toggles. It is better to have smaller and more focused snapshots. More info on this approach can be found in this article [Effective Snapshot Testing](https://kentcdodds.com/blog/effective-snapshot-testing).

An even better example of asserting a headline is on the page

```js
it('I can see the headline', () => {
  const h1El = document.querySelector('h1');

  expect(h1El).toBeInTheDocument(); // check the headline element is in the document
  expect(h1El.getAttribute('id')).toBe('content'); // check for the id attribute
  expect(h1El.getAttribute('tabindex')).toBe('-1'); // check for the tabindex attribute

>   expect(h1El).toBeTruthy(); // check there is some text inside the element

Reviewer 2: Should this be `expect(h1El.textContent).toBeTruthy()` as per my previous comment?

A docblock pragma is a specially-formatted comment at the top of a test file. We can use docblock pragmas to alter the environment that tests run in.

## What is a snapshot?

> This is a feature provided by Jest. It's not a snapshot of the graphical UI but a snapshot of a result that is typically returned from a function. This could even be HTML returned from DOM query selector e.g.

Reviewer: Just a change of emphasis suggestion, won't we be using `toMatchSnapshot` to match html most of the time?
```suggestion
This is a feature provided by Jest. It's not a snapshot of the graphical UI but a snapshot of a result that is typically returned from a function. This is often the HTML returned from DOM query selector e.g.
```
